linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/external/yarn/YarnResourceRequester.java

```diff
@@ -0,0 +1,426 @@
+
+    } else if ("fairScheduler".equals(schedulerType)) {
+      JsonNode childQueues =
+          getChildQueues(resp.path("scheduler").path("schedulerInfo").path("rootQueue"));
+      Optional<JsonNode> queue = getQueue(childQueues, realQueueName);
+      if (!queue.isPresent()) {
+        logger.debug(
+            "cannot find any information about queue " + queueName + ", response: " + resp);
+        throw new RMWarnException(
+            YARN_NOT_EXISTS_QUEUE.getErrorCode(),
+            MessageFormat.format(YARN_NOT_EXISTS_QUEUE.getErrorDesc(), queueName));
+      }
+      Optional<JsonNode> maxResources = queue.map(node -> node.path("maxResources"));
+      Optional<JsonNode> usedResources = queue.map(node -> node.path("usedResources"));
+      return Pair.of(
+          getYarnResource(maxResources, queueName).get(),
+          getYarnResource(usedResources, queueName).get());
+
+    } else {
+      logger.debug(
+          "only support fairScheduler or capacityScheduler, schedulerType: "
+              + schedulerType
+              + ", response: "
+              + resp);
+      throw new RMWarnException(
+          ONLY_SUPPORT_FAIRORCAPA.getErrorCode(),
+          MessageFormat.format(ONLY_SUPPORT_FAIRORCAPA.getErrorDesc(), schedulerType));
+    }
+  }
+
+  public static Optional<YarnResource> getYarnResource(
+      Optional<JsonNode> jsonNode, String queueName) {
+    if (jsonNode.isPresent()) {
+      JsonNode r = jsonNode.get();
+      return Optional.of(
+          new YarnResource(
+              r.get("memory").asLong() * 1024L * 1024L, r.get("vCores").asInt(), 0, queueName, ""));
+    }
+    return Optional.empty();
+  }
+
+  @Override
+  public List<ExternalAppInfo> requestAppInfo(
+      ExternalResourceIdentifier identifier, ExternalResourceProvider provider) {
+    String rmWebHaAddress = (String) provider.getConfigMap().get("rmWebAddress");
+    String rmWebAddress = getAndUpdateActiveRmWebAddress(rmWebHaAddress);
+    String queueName = ((YarnResourceIdentifier) identifier).getQueueName();
+
+    String realQueueName = "root." + queueName;
+
+    JsonNode resp = getResponseByUrl("apps", rmWebAddress).path("apps").path("app");
+    if (resp.isMissingNode()) {
+      return new ArrayList<>();
+    }
+    ArrayNode appArray = (ArrayNode) resp;
+
+    List<ExternalAppInfo> appInfoList = new ArrayList<>();
+    Iterator<JsonNode> iterator = appArray.elements();
+    while (iterator.hasNext()) {
+      JsonNode app = iterator.next();
+      String yarnQueueName = app.get("queue").asText();
+      String state = app.get("state").asText();
+      if (yarnQueueName.equals(realQueueName)
+          && (state.equals("RUNNING") || state.equals("ACCEPTED"))) {
+        String id = app.get("id").asText();
+        String user = app.get("user").asText();
+        String applicationType = app.get("applicationType").asText();
+        Optional<YarnResource> yarnResource = getYarnResource(Optional.ofNullable(app), queueName);
```

Comment:
does this `getYarnResource `  work ?

Reply:
Method getYarnResource works ,I had trace this method using arthas when  I submit spark task.

Reply:
Thanks